### DISCUSSION TITLE
Transcoda: New omr model

### DISCUSSION THREAD

**Discussion by aicelen** [liebharc/homr#95] [General]
I recently found [transcoda](https://github.com/btrkeks/transcoda) and wanted to share this with you.
It uses a model in a similar size (59M Parameters) and is pretty interesting. It inputs the whole image at a fixed resolution (I assume that this will hurt performance when using bigger images) and directly outputs in kern notation. The problem is that it sometime makes mistakes which break the humdrum viewer (https://verovio.humdrum.org/). When I use the converter to convert to mxl the generated files is corrupted.
It performs better than homr on the single staff image in the readme (probably because of the low resolution).
Also it is trained only on synthetic data (300k images/staffs?) rendered with verovio (like grandstaff) dataset.

**Comment by liebharc** [comment]
Nice! Thanks for sharing that. I didn't read the paper in detail yet but skimmed through it. Here are some notes.

Disclaimer first: I really only spent like 30min looking at the paper and data and I might be missing things. 

- That's the first contribution to OMR from Heinrich Heine University Düsseldorf which I'm aware of (or I just forgot) - nice to have a new group working on OMR
- The model seems to be close to SMT - at least from top of my head, I haven't double-checked the SMT paper
- The main contribution of this paper seems to be the training on purely synthetic data sets
- They are using a constrained decoding. I think we had something like that in the past, but newer versions of homr don't have it anymore - mainly because the benefit was little.
- They also use an end-to-end approach like SMT - and in contrast to our 2-stage approach
- Another model trained on humdrum kern, that's just something we discussed in https://github.com/liebharc/homr/discussions/74 . They remove the ambiguity issues I mentioned by making sure that their synthetic generation follows certain rules.

The idea of synthetic data sets is something which I discussed with people (in real life) before. The obvious benefit is that it's rather straight forward, and you can generate a lot of data. The risk is that you create a lot of data which is essentially the same, so that it lacks variance. E.g. certain aspects can be completely absent. I would think that we could certainly reuse the ideas of data generation and data augmentation to extend/replace/improve our training pipeline. 

Regarding the end-to-end approach: My intuition would be that this works well for scanned sheet music. homr was created to work with images from smartphone cameras where you have really diverse backgrounds and noise. I would feel like in these scenarios the end-to-end approach would likely fail to filter out the background and noise - and that it will be hard to extend the training of the end-to-end models to handle these cases. "Hard" obviously doesn't mean that it never happens. And "intuition" means that I have run no tests to confirm what I'm saying :). As a practical example: There was a case where homr was used on an image with sheet music, but also with a cat, and it successfully ignored that cat and translated the sheet music (although if I write that it feels sad that it ignored the cat). That's the scenarios where end-to-end models might struggle. There also is https://github.com/RobbenWijanathan/adversarial-homr which seems to work on checking the robustness of homr.

And perhaps one should write a research paper on homr at some point. It would help so that other publications might directly compare their approach with ours.

> It performs better than homr on the single staff image in the readme

I'm not much surprised by that. If the example image comes from the same data generator as their training data, then their model will naturally perform better on it - even if the particualar image isn't in the training set. The part which makes me a bit sad, is that the image really doesn't even look that hard and homr should deal well with it. But well, OMR is a hard topic. The better comparison would be to compare it one the Polish scan benchmark.

Update: I just started wondering if the reason for the better performance of Transcoda over SMT might not be the amount of data, but just simply be caused that their synthetic data doesn't create the ambiguities in kern representation. But perhaps I'm just too focused on one particular aspect today.

**Comment by liebharc** [comment]
I had some time to go through the paper in more detail and found that I misunderstood a couple of things as I skimmed through it earlier.

The paper makes a lot of sense. Their thoughts around the training pipeline match a lot with what homr does: Take different data sources, normalize them. They go beyond what we do in several ways:

- More datasets: PDMX and MuseTrainer 
- Better augmentation, e.g. notation injection could be really important

Obviously their end-to-end approach is different from our 2-stage approach. Also, the details of how they use their convnext encoder differ from ours.

From a first benchmark (https://github.com/liebharc/homr/pull/100, take them with a grain of salt) transcoda and homr are showing about the same performance on the polish-scores dataset (transcoda has the better score, but only because it fails on some hard examples which artificially lower the score in its favor). That is impressive as homr obviously has had more time to implement optimizations.

We could likely benefit from their data pipeline, but then need to invest additional work (likely considerable) to keep the current feature set of homr (orchestra pieces with multiple voices/instruments is likely one of them).

**Comment by liebharc** [comment]
This also just came out:

https://grfia.dlsi.ua.es/musvit/
https://arxiv.org/pdf/2606.31811

It's the University of Alicante who are very active in the OMR field and have a lot of experience

**Reply by aicelen** [reply]
This is really interesting.
Do you think it could be worth it to replace our current vision encoder with this one? The input shape is with 1024x1024 roughly 4x larger - if we would continue to only use single/grand staff inputs, homr would probably become way slower.

**Reply by liebharc** [reply]
Not right no I think. I did a quick test run - while the train on 512x512 images and 1024x1024 you can use various sizes and their patch size of 16 matches ours already. For non-square images they even mention interpolation as alternative to padding here: https://huggingface.co/PRAIG/musvit-light . I let it run for 6 epochs and SER was a bit better (15% vs 17%) but this early in the training that doesn't mean much. The legal part is more restricting: CC BY-NC-SA 4.0 is incompatible with AGPL as it prohibits commercial use. So we can't use their code and weights - apart from tests which don't make it into the code base or a trained model. 

Overall I would wait and see if there are follow-up papers. The benchmark results they published are hard for me to compare. Like the polish scores numbers look really good, but without having seen the details and since I don't know any of the other models they use as reference, it doesn't help me much yet.
